### DISCUSSION TITLE
Adding feature to lb4 cli

### DISCUSSION THREAD

**Discussion by aaqilniz** [loopbackio/loopback-next#8686] [Q&A]
Hi. I have a very specific scenario that I would like to add to lb4 CLI. Is it possible for me to add this scenario to the CLI package to the loopback-next's repo? Or is there any way I can implement this particular scenario? 

I actually need a CLI command that would take in a JSON formatted SQL query. The command needs to create a new route by creating a new controller or updating the existing one. Once that route is hit by the user (through browser or HTTP client), it must return all the data generated by that query.

Thanks.

**Comment by achrinza** [comment]
Thanks for starting a discussion on this; There's some points that come into mind:

> JSON formatted SQL query.

What would be the final format of this JSON SQL query? Is it based on an existing standard/specification?

> or updating the existing one.

Creating a new controller would be possible, but updating is not something that's been done with the CLI; The reason being that a LoopBack 4 application can take many shapes, and we'd want to reduce the number of assumptions the CLI makes about the LB4 app. Updating an existing controller requires reading existing project files, which can break easily if the developer customizes it (i.e. How much customization is acceptable before the CLI breaks?).

>  it must return all the data generated by that query.

I think it definitely is a quality-of-life feature if we the LB4 CLI can generate the skeleton code needed to return the result of a complex SQL query.

However, do we want to generate using the existing artifacts, or do we want to create a new artifact for this feature?

The `@loopback/rest-crud` package is an example of the latter - The tradeoff is extensibility over simplicity. This may be a good model to follow since SQL queries can be powerful, and can be used to manipulate the data returned without needing to write additional TypeScript. However, it also means that the developer is now restricted to the SQL flavour of the database engine, which may not be ideal.

**Reply by aaqilniz** [reply]
I am really grateful for the response.

I said JSON but I think it'll be a string with a predefined structure. For example;

dsource|dmodel|method|params....|sqlQuery

We can't update the controllers with the CLI but can create multiple new controllers for the same model, right?

Is it possible to create a new controller with a custom URI path? 

When we create a controller it can either be CRUD or Empty controller, how about a third option with Custom and it asks for data source, model, HTTP method and so on.
